### ISSUE TITLE
View updates should be generated by sstables introduced via nodetool refresh

### ISSUE THREAD

**Issue by duarten** [scylladb/scylla#4047]
The sstables that are placed in the `/upload` directory are introduced in the system by `nodetool refresh`. Currently, we don't generate view updates for them, but we should, in a fashion similar to what we do for streaming sstables.

**Comment by duarten** [comment]
@glommer sstables in /upload will now have view updates generated for them.

**Comment by glommer** [comment]
thank you that's very kind of you
3.1 ?

**Comment by duarten** [comment]
Yep!

**Comment by glommer** [comment]
I just got to reading the code for this while searching for another issue

There is a comment now in `flush_upload_dir` that says the following:

```
 // If view updates need to be generated, leave them in upload/ directory, so they will be treated as staging sstables.
```
The code seems to imply that we move the base sstables to the table directory, leave the sstables in the upload directory untouched, and return the future.

This seems really really not good to me (assuming correct understanding) for the following reason:
 if we return, nodetool refresh will return and the user at this point may be ready to upload more sstables. However since there are SSTables left in the upload directory, the user may override the old SSTables.

Checking to see if the directory is finally empty is non-scriptable and also an incredibly surprising behavior. I consider this a regression (if my understanding is correct).

We should create a dedicated staging directory per upload attempt (so they don't clash), and leave the SSTables there to make the base `upload` usable again.

I am reopening this since the way it was solved (once more, if I understand this correctly) this is a data corruption waiting to happen.

If you think I am mistaken in my analysis just close again and explain. 

Thanks!

**Comment by psarna** [comment]
You're not mistaken, I just regarded uploading through `/upload` as coming from one corresponding data directory we migrate from, so I imagined it as being either one-shot upload of all sstables in it, or, even if done partially, a continuous upload from one directory, which by POSIX rules doesn't have two files with the same name.

Now, creating per-upload-session upload directory is an option, but I have two questions/remarks:
1. Is it *ever* the case that we `mount` the upload directory with a device that stored data from a machine we migrate from? Because then creating other directories in the same root will not behave as expected
2. It's doable, but includes adding yet another mechanism of handling failures - when an sstable is halfway through being transferred from `/upload` to `/upload-session-X`, which is its designated staging directory.
3. If we create these per-upload-session dirs in root, we can potentially see an explosion of unrelated dirs in a parent directory where data otherwise lives, so it sounds bad. I suppose we can have yet another special directory called `/upload-per-session`, `staging-per-upload-session` or whatnot to serve the purpose. The drawback is that we end up with one more cryptic directory that very few people know the purpose of, even if it's documented.
4. How about returning the future only after the updates are generated from the sstables? That will considerably delay the response to the client, but will on the other hand make it clear that uploading sstables is actually over - not only are they live in the system, but also all view updates are properly generated from them.
5. To supplement `4.`, if we made the update generation part optional, we could have two paths depending on what the migration procedure is: if we just supplement the cluster with a small number of sstables, we can afford waiting for the future to complete. If we use `/upload` to migrate everything, we can just upload data first and recreate views later to avoid the directory mess.
Thoughts?

**Comment by glommer** [comment]
Hi @psarna my apologies. I wasn't notified on this issue (or I was, and got lost in the noise).

Let me address your comments:

```
a continuous upload from one directory, which by POSIX rules doesn't have two files with the same name.
```

That is exactly the problem =) We are not defending here against a Scylla bug with duplicate tables.
We are defending against the fact that if the user now `cp` a new set of SSTables and one of them has the same name, the new file will take its place, the old one will be gone. This will lead to data loss.

About your questions / remarks:

1. Since we use hard links, we should assume there is no extra mountpoint and everything is in the same filesystem

2. I understand, but I see no better option (aside from point 4.)

3. Those directories can all be inside `upload/` For instance , `upload/ongoing-view-updates-xxx`

4. We can do that too. Do keep in mind that operations that take a long time are annoying during recovery. View build is supposed to be asynchronous, and this would essentially make them synchronous during upload. I don't like this option.

5. seems too complex.

`nodetool refresh` is by far the most used path for disaster recovery of all sorts. I am tagging this issue now as high prio.

**Comment by glommer** [comment]
@slivne to make sure you are aware: I consider this a 3.1 blocker.

**Comment by glommer** [comment]
@eliransin
